config/ols-rbac.yaml

```diff
@@ -0,0 +1,63 @@
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: ols-proxy
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: ols-proxy
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: ols-proxy
+subjects:
+- kind: ServiceAccount
+  name: ols-proxy
+  namespace: ols
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: ols-proxy
+rules:
+- apiGroups: ["authentication.k8s.io"]
+  resources:
+  - tokenreviews
+  verbs: ["create"]
+- apiGroups: ["authorization.k8s.io"]
+  resources:
+  - subjectaccessreviews
+  verbs: ["create"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: ols-proxy-backend-register
+rules:
+- nonResourceURLs:
+    - "/docs"
+    - "/ols"
+    - "/base_llm_completion"
+    - "/feedback"
+  verbs:
+    - get
+    - create
+    - delete
+    - list
+    - update
+    - patch
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: ols-proxy-backend-register
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: ols-proxy-backend-register
+subjects:
+- kind: ServiceAccount
+  name: default
```

Comment:
we should probably never use the default SA for anything, create a new one.

but also i'm not clear why the OLS SA needs this rolebinding?

Reply:
It might be good to start with a an SA with read-only role (e.g. cluster-reader) for future querying of the system during  contextualization.

Reply:
cluster context retrieval is not something we are looking at yet, if/when we get there it will probably need to be its own SA entirely so we can separate permissions needed to interrogate the cluster(it may even be an entirely separate process), from permissions OLS needs for running the client->OLS->LLM flows and validating requests.

So let's keep the SA permission focused on that latter bit for now.

Reply:
OLS SA doesn't require the above rolebinding, its just an SA from client side. Create a new one for demonstrating the `client-test` example.

Reply:
OLS SA doesn't require this role binding. This is for client and can be in the client's test namespace, but looks like it has to use only `default` service account. This seems to be mandatory on `kube-rbac-proxy` end. I did an experiment by deploying client in a different namespace and created a different service account, but kube-proxy automatically looks for `default` service account no matter what SA we create in client's namespace for authorization. Relevant failure logs here:

```
I1115 02:18:47.899587       1 request.go:1172] Request Body: {"kind":"SubjectAccessReview","apiVersion":"authorization.k8s.io/v1","metadata":{"creationTimestamp":null},"spec":{"nonResourceAttributes":{"path":"/docs","verb":"get"},"user":"system:serviceaccount:ols-test:default","groups":["system:serviceaccounts","system:serviceaccounts:ols-test","system:authenticated"],"extra":{"authentication.kubernetes.io/pod-name":["ols-client-7cbl5"],"authentication.kubernetes.io/pod-uid":["b1227ff6-0b69-4a43-87f1-678e405ad560"]},"uid":"7a2f0494-47cf-429c-baf0-2b954daba0e0"},"status":{"allowed":false}}


I1115 02:18:47.901768       1 request.go:1172] Response Body: {"kind":"SubjectAccessReview","apiVersion":"authorization.k8s.io/v1","metadata":{"creationTimestamp":null,"managedFields":[{"manager":"kube-rbac-proxy","operation":"Update","apiVersion":"authorization.k8s.io/v1","time":"2023-11-15T02:18:47Z","fieldsType":"FieldsV1","fieldsV1":{"f:spec":{"f:extra":{".":{},"f:authentication.kubernetes.io/pod-name":{},"f:authentication.kubernetes.io/pod-uid":{}},"f:groups":{},"f:nonResourceAttributes":{".":{},"f:path":{},"f:verb":{}},"f:uid":{},"f:user":{}}}}]},"spec":{"nonResourceAttributes":{"path":"/docs","verb":"get"},"user":"system:serviceaccount:ols-test:default","groups":["system:serviceaccounts","system:serviceaccounts:ols-test","system:authenticated"],"extra":{"authentication.kubernetes.io/pod-name":["ols-client-7cbl5"],"authentication.kubernetes.io/pod-uid":["b1227ff6-0b69-4a43-87f1-678e405ad560"]},"uid":"7a2f0494-47cf-429c-baf0-2b954daba0e0"},"status":{"allowed":false}}
I1115 02:18:47.901872       1 auth.go:97] Forbidden (user=system:serviceaccount:ols-test:default, verb=get, resource=, subresource=). Reason: "".
```
The same request works fine when we have `default` SA bound. It always looks for `default` SA and fails no matter what service account we have on the client side. So, only default SA should be bound with client-side role bindings.

I see others using different SA [here](https://github.com/redhat-appstudio/sprayproxy/blob/main/config/sprayproxy-rbac.yaml#L61) and they do use a different [image](https://github.com/redhat-appstudio/sprayproxy/blob/main/config/sprayproxy.yaml#L48). I am not sure if they are able to make requests using a token or not, but I tried it in my local and it failed for me with the same error above.

Let me know if this is something not acceptable and if we need to raise a issue with `kube-rbac-proxy` team on this.

Reply:
If i'm understanding what you're saying correctly, regardless of what SA token the client provides, the rbac proxy is attempting to check if the default SA from the client's NS has the right permissions?


Keep in mind the end-goal here is going to be that a user can make requests to OLS using their own creds/token.  So the idea that kube-rbac is falling back to trying to use the default SA (from what NS?  it seems to have picked the client NS somehow, but how did it do that?) is going to be a problem for how we intend to use this. 

It must be a configuration issue.  Try reaching out to Jeff Cantrill.  Tag me in the conversation

Reply:
Fixed it as per our discussion offline: https://redhat-internal.slack.com/archives/C05EG043LER/p1700070319423039